test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeMessageForwardTest.java

```diff
@@ -0,0 +1,127 @@
+    private static String INSERT_QUERY = String.format("INSERT INTO %s.%s(pk) VALUES (?)", KEYSPACE, TABLE);
+    private static String CHECK_QUERY = String.format("SELECT pk FROM %s.%s WHERE pk = ?", KEYSPACE, TABLE);
+
+    private boolean checkClusterUp(UpgradeableCluster cluster, int coordId)
+    {
+        NodeToolResult result;
+        result = cluster.get(coordId).nodetoolResult("ring");
+
+        // Must have an Up line for each node
+        long upCount = Arrays.stream(result.getStdout().split("\\r?\\n")).filter(line -> line.contains(" Up ")).count();
+        if (upCount < cluster.size())
+        {
+            logger.info("Only {}/{} are up.", upCount, cluster.size());
+            return false;
+        }
+
+        return true;
+    }
+
+    private void writeReadTest(UpgradeableCluster cluster)
+    {
+        // Coordinate a write from each node and then check present on all replicas
+        int readKey = nextKey;
+        for (int coordId = 1; coordId <= cluster.size(); coordId++)
+        {
+            final int checkNodeId = coordId;
+            Awaitility.await("Cluster Up").atMost(1, TimeUnit.MINUTES).until(() -> checkClusterUp(cluster, checkNodeId));
+
+            logger.info("Coordinating CL.ALL Insert from node{} ", coordId);
+            cluster.get(coordId).coordinator().execute(INSERT_QUERY, ConsistencyLevel.ALL, nextKey++);
+        }
+
+        for (int coordId = 1; coordId <= cluster.size(); coordId++)
+        {
+            for (int nodeId = 1; nodeId <= cluster.size(); nodeId++) {
+                Object[][] results = cluster.get(nodeId).executeInternal(CHECK_QUERY, readKey);
+                assertRows(results, row(readKey));
+            }
+            readKey++;
+        }
+    }
+
+    /* Verify that messages sent with sendToHintedReplicas to non-local DCs
+     * are forwarded on to the hosts there.
+     *
+     * 1) creates a mixed cluster with multiple datacenters and a keyspace
+     *    configured to write to all replicas in the datacenter
+     * 2) check the original single-version cluster by issuing an INSERT
+     *    mutation from a coordinator on each node, then check that value
+     *    has locally been written to each of the nodes.
+     * 3) Upgrade nodes one at a time, rechecking that all writes are forwarded.
+     */
+    @Test
+    public void checkWritesForwardedToOtherDcTest() throws Throwable
+    {
+        int numDCs = 2;
+        int nodesPerDc = 3;
+        String ntsArgs = IntStream.range(1, numDCs + 1)
+                                  .mapToObj(dc -> String.format("'datacenter%d' : %d", dc, nodesPerDc))
+                                  .collect(Collectors.joining(","));
+
+        new TestCase()
+        .withConfig(c -> c.with(Feature.GOSSIP, Feature.NETWORK).set("request_timeout_in_ms", 30000))
+        .withBuilder(b -> b.withRacks(numDCs, 1, nodesPerDc))
+        .nodes(numDCs * nodesPerDc)
+        .upgrade(Versions.Major.v30, Versions.Major.v4)
```

Comment:
nit: Should we test 3.11 at all, or is the serialization apparatus basically the same in 3.11 as it is in 3.0?

Reply:
Serialization looked the same when I checked 3.11.  Not sure it's worth running every release, but I can test it once off.

Reply:
Confirmed running with `.upgrade(Versions.Major.v3X, Versions.Major.v4)` works. 

Before this change there would be value in testing `nodesPerDC=2;` and a larger value to exercise different paths through `sameMessageId`, however the updated implementation doesn't special case single node forwarding.